pkg/destroy/aws/aws.go

```diff
@@ -77,1510 +63,1205 @@ func (o *ClusterUninstaller) validate() error {
 	return nil
 }
 
-// populateDeleteFuncs is the list of functions that will be launched as goroutines
-func populateDeleteFuncs(funcs map[string]deleteFunc) {
-	funcs["deleteVPCs"] = deleteVPCs
-	funcs["deleteEIPs"] = deleteEIPs
-	funcs["deleteNATGateways"] = deleteNATGateways
-	funcs["deleteInstances"] = deleteInstances
-	funcs["deleteIAMresources"] = deleteIAMresources
-	funcs["deleteSecurityGroups"] = deleteSecurityGroups
-	funcs["deleteInternetGateways"] = deleteInternetGateways
-	funcs["deleteSubnets"] = deleteSubnets
-	funcs["deleteS3Buckets"] = deleteS3Buckets
-	funcs["deleteRoute53"] = deleteRoute53
-	funcs["deletePVs"] = deletePVs
-	funcs["deleteUsers"] = deleteUsers
-}
-
 // Run is the entrypoint to start the uninstall process
 func (o *ClusterUninstaller) Run() error {
 	err := o.validate()
 	if err != nil {
 		return err
 	}
 
-	deleteFuncs := map[string]deleteFunc{}
-	populateDeleteFuncs(deleteFuncs)
-	returnChannel := make(chan string)
+	awsConfig := &aws.Config{Region: aws.String(o.Region)}
 
-	awsSession, err := getAWSSession(o.Region)
+	// Relying on appropriate AWS ENV vars (eg AWS_PROFILE, AWS_ACCESS_KEY_ID, etc)
+	awsSession, err := session.NewSession(awsConfig)
 	if err != nil {
 		return err
 	}
 
-	// launch goroutines
-	goroutines := 0
-	for name, function := range deleteFuncs {
-		for _, filter := range o.Filters {
-			go deleteRunner(name, function, awsSession, filter, o.ClusterName, o.Logger, returnChannel)
-			goroutines++
+	tagClients := []*resourcegroupstaggingapi.ResourceGroupsTaggingAPI{
+		resourcegroupstaggingapi.New(awsSession),
+	}
+	tagClientNames := map[*resourcegroupstaggingapi.ResourceGroupsTaggingAPI]string{
+		tagClients[0]: o.Region,
+	}
+	if o.Region != "us-east-1" {
+		tagClient := resourcegroupstaggingapi.New(
+			awsSession, aws.NewConfig().WithRegion("us-east-1"),
+		)
+		tagClients = append(tagClients, tagClient)
+		tagClientNames[tagClient] = "us-east-1"
+	}
+
+	deleted := map[string]struct{}{}
+	iamClient := iam.New(awsSession)
+	iamRoleSearch := &iamRoleSearch{
+		client:  iamClient,
+		filters: o.Filters,
+		logger:  o.Logger,
+	}
+	iamUserSearch := &iamUserSearch{
+		client:  iamClient,
+		filters: o.Filters,
+		logger:  o.Logger,
+	}
+
+	var loopError error
+	for len(tagClients) > 0 || loopError != nil {
+		loopError = nil
+		nextTagClients := tagClients[:0]
+		for _, tagClient := range tagClients {
+			matched := false
+			for _, filter := range o.Filters {
+				o.Logger.Debugf("search for and delete matching resources by tag in %s matching %#+v", tagClientNames[tagClient], filter)
+				tagFilters := make([]*resourcegroupstaggingapi.TagFilter, 0, len(filter))
```

Comment:
Why are we doing this in hot path, when `o.Filters` is global and same for all the runs.

Reply:
Because I expect this loop to be limited by AWS API round-trips vs. local memory allocation for small maps, and I hadn't bothered to set up a `[][]*resourcegroupstaggingapi.TagFilter` to iterate over.  Would you like a variable like that?

Reply:
its fine as-is for now.